src/security/body_parse/header.cpp

```diff
@@ -0,0 +1,520 @@
+#include "header.h"
+
+#include <algorithm>
+#include <iterator>
+#include <string_view>
+
+#include "../decode.h"
+#include "chain_is.h"
+#include "generator.h"
+
+using namespace std::literals;
+
+namespace dnsec = datadog::nginx::security;
+
+namespace {
+
+inline bool equals_ci(std::string_view a, std::string_view lc_b) {
+  if (a.size() != lc_b.size()) {
+    return false;
+  }
+  for (std::size_t i = 0; i < a.size(); i++) {
+    if (std::tolower(a[i]) != lc_b[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+inline std::string to_lc(std::string_view sv) {
+  std::string result;
+  result.reserve(sv.size());
+  std::transform(sv.begin(), sv.end(), std::back_inserter(result),
+                 [](unsigned char c) { return std::tolower(c); });
+  return result;
+}
+
+void consume_ows(std::string_view &sv) {
+  while (!sv.empty() && (sv.front() == ' ' || sv.front() == '\t')) {
+    sv.remove_prefix(1);
+  }
+}
+
+/*
+ * Consumes a token according to RFC 9110, section 5.6.2.
+ *
+ * https://httpwg.org/specs/rfc9110.html#rfc.section.5.6.2
+ *   token          = 1*tchar
+ *   tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
+ *                  / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
+ *                  / DIGIT / ALPHA
+ *                  ; any VCHAR, except delimiters
+ *
+ * For multipart/form-data, RFC 9110 refers to RFC 7578 ("Returning Values from
+ * Forms: multipart/form-data"), which in refers to RFC 2183 ("The
+ * Content-Disposition Header Field"), which in turn defines their tokens like
+ * this:
+ *
+ * https://datatracker.ietf.org/doc/html/rfc2045 (by reference to RFC 822)
+ *   token      := 1*<any (US-ASCII) CHAR except SPACE, CTLs,
+ *                    or tspecials>
+ *
+ *   tspecials :=  "(" / ")" / "<" / ">" / "@" /
+ *                 "," / ";" / ":" / "\" / <">
+ *                 "/" / "[" / "]" / "?" / "="
+ *                 ; Must be in quoted-string,
+ *                 ; to use within parameter values
+ *
+ * This is both more permissive (allows {}) and more restrictive (forbids
+ * characters outside ASCII).
+ */
+std::optional<std::string_view> consume_wg_token(std::string_view &sv) {
+  static constexpr std::string_view tchar =
+      "abcdefghijklmnopqrstuvwxyz"
+      "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+      "0123456789!#$%&'*+-.^_`|~";
+  auto end = sv.find_first_not_of(tchar);
+  if (end == 0) {
+    return std::nullopt;
+  }
+  if (end == std::string_view::npos) {
+    end = sv.size();
+  }
```

Comment:
This is not needed, when the `count` passed to `substr` is `npos`, the substring will be from `start` until the end of the string.

Reply:
Yes, but `end` is also used on the `remove_prefix` call, which doesn't accept npos

Reply:
Ah yes, I missed that entirely.